### ISSUE TITLE
Order of layers in model history

### ISSUE THREAD

**Issue by tlasmanbotch** [johnmarktaylor91/torchlens#14]
Hello -

Thank you for the fantastic package! I've been using this with a few CNN architectures (e.g., AlexNet, ResNet50) and had a question about the module hierarchy reported in the model history.

Specifically, reporting the modules from ResNet50 (which has nested, sequential blocks) appears to take a different order than what is reported by running model.eval(). I'm showing an example below.

Within the model evaluation, batch normalization is interleaved in between convolutions and ReLUs:

![Screen Shot 2023-09-13 at 11 26 31 AM](https://github.com/johnmarktaylor91/torchlens/assets/42591108/e77f5113-2d39-44ff-92c8-9bed0f578129)

But in the module hierarchy of torchlens, the layers seem to be reported out of order:

![Screen Shot 2023-09-13 at 11 27 25 AM](https://github.com/johnmarktaylor91/torchlens/assets/42591108/c794778b-8b92-450d-b1e4-4c098a9cb400)

I have two questions here:
1.  Would this ordering have any effect on how the activations are retrieved? 
2.  Is run-time expected to be high for these nested networks? For a single forward pass of one image on CPU, it is taking ~40 seconds even when not logging any layers (compared w/ ~5s when using model.forward()). 

Thanks in advance for your insight!

**Comment by johnmarktaylor91** [comment]
Hey there, thanks so much for the kind words and for the feedback--this is super helpful for me so I know where to improve the package and documentation :] 

Regarding layer ordering:
1. Just to clarify one slightly confusing thing, TorchLens make a distinction between "layers" (any PyTorch tensor operation) and "modules" (the PyTorch data structure). TorchLens treats "layers" as primary, with modules being the "containers" in which layers/operations are executed. This is because there are some PyTorch operations that aren't modules, and so "layers" seemed like the most natural generic term to use (at least in my brain). 

2. When you call print on the ModelHistory object (i.e., the output of log_forward_pass), it separately lists both the layers and the modules. Both are listed based on execution order, with the slightly confusing caveat that "buffer layers" (i.e., non-trainable tensors that are stored along with the model; batchnorm layers use these) are counted as being "executed" first. This causes the batchnorm modules to appear first in the module printout. 

3. By contrast, if you just call print on a regular PyTorch model object (as you did in your post), the modules are listed NOT based on execution order during the forward pass, but just based on the order in which they were defined when the model was initialized. This can make the printout potentially misleading in some cases, because there's no requirement that these line up. For example, in the ResNet50 printout you show, the ReLU module is printed out *after* all the conv and batchnorm layers, even though the ReLU operation is interleaved between each conv-batchnorm repetition. I should note that TorchLens saves each execution of the ReLU module: layer1.0.relu:1 is the first pass, layer1.0.relu:2 is the second pass, and so on. 

So long story short, the discrepancy here is just due to a difference in the printout order: for the default PyTorch printout modules are listed based on the order they're initialized in the code, whereas TorchLens lists them based on execution order during the forward pass. This is a purely cosmetic difference that shouldn't affect how the activations are saved. This is where I hoped having the TorchLens model visualization could be useful, by giving you a flow chart that tells you exactly how the layers and modules are organized in the network (since often a single sequential ordering can't convey the branching structure). 


Regarding run-time: 
1. The initial call to log_forward_pass for TorchLens can indeed be pretty slow, since it's tracking a bunch of metadata and doing a bunch of processing of the model's computational graph. This isn't due to the nesting per se, but it certainly gets slower for more complex networks in general. 

2. For this reason, I recently added a new function, save_new_activations, that's meant to speed up the process of saving activations. The CoLab gives an example of this, but basically you call log_forward_pass on the model (with any input you want), and then the resulting ModelHistory output object has a method called save_new_activations that allows you to pass in a new input and get new activations without having to re-process all the metadata. This speeds things up considerably (see attached screenshot). So in general, the current workflow is that there's just one slow step for each model to log all the metadata, but then saving new activations for that model should be pretty fast. 

3. In the future I'm thinking of adding another function to speed up saving activations even more (i.e., that *only* saves activations without saving any metadata at all)--I can bump this up the priority list if speed becomes a bottleneck for folks. 

<img width="546" alt="image" src="https://github.com/johnmarktaylor91/torchlens/assets/16569815/06d50aa7-3fe4-4d2f-833e-bde2304f1c80">

Sorry for the verbose response, but lemme know if that answers your questions and if there's anything else I can clarify or improve :]

**Comment by tlasmanbotch** [comment]
Thanks for the response! All of these thoughts answer my questions and are helpful in understanding the design of the package. Closing the issue.

**Comment by johnmarktaylor91** [comment]
Glad I could help, thanks for the feedback!
